crates/wasmtime/src/store/data.rs

```diff
@@ -0,0 +1,256 @@
+where
+    StoreData: Index<I>,
+{
+    type Output = <StoreData as Index<I>>::Output;
+
+    #[inline]
+    fn index(&self, index: I) -> &Self::Output {
+        self.store_data().index(index)
+    }
+}
+
+#[repr(transparent)]
+pub struct Stored<T> {
+    // See documentation below on `INDEX_BITS` for how this is interpreted.
+    bits: u64,
+    _marker: marker::PhantomData<fn() -> T>,
+}
+
+// This is the maximal number of bits that the index of an item within a store
+// can take up. As this is set to 24 that allows for 16 million items. Note
+// that this is not a limit on something like the number of functions within an
+// instance, only a limit on the number of externally referenced items in a
+// Store. For example this is more equivalent to exported functions of a module
+// rather than functions themselves.
+//
+// The reason for this limitation is that we want `Stored<T>` to fit into a
+// 64-bit value (for the C API). This 64-bit value gives us limited, well, uh,
+// bits, to work with. We need to pack both a "store id" as well as an index
+// within the store into those 64 bits. Given that there's no implementation of
+// recycling store IDs at this time it also means that the number of bits
+// allocated to the store id represents the maximal number of stores that a
+// process can create for its entire lifetime.
+//
+// These factors led to the choice of bits here for this. This can be moved
+// around a bit, but the hope is that this is good enough for all practical
+// users.
+//
+// The choice of 24 means that the limitations of wasmtime are:
+//
+// * 24 bits for the index, meaning 16 million items maximum. As noted above
+//   this is 16 million *host references* to wasm items, so this is akin to
+//   creating 16 million instances within one store or creating an instance
+//   that has 16 million exported function. If 24 bits isn't enough then we
+//   may need to look into compile-time options to change this perhaps.
+//
+// * 40 bits for the store id. This is a whole lot more bits than the index,
+//   but intentionally so. As the maximal number of stores for the entire
+//   process that's far more limiting than the number of items within a store
+//   (which are typically drastically lower than 16 million and/or limited via
+//   other means, e.g. wasm module validation, instance limits, etc).
+//
+// So all-in-all we try to maximize the number of store bits without placing
+// too many restrictions on the number of items within a store. Using 40
+// bits practically means that if you create 10k stores a second your program
+// can run for ~3.5 years. Hopefully that's enough?
+//
+// If we didn't need to be clever in the C API and returned structs-by-value
+// instead of returning 64-bit integers then we could just change this to a
+// u64/usize pair which would solve all of these problems. Hopefully, though,
+// no one will ever run into these limits...
+const INDEX_BITS: usize = 24;
```

Comment:
After writing all this and thinking more about it, I think another possible solution could be:

* In Rust `Stored<T>` is 128 bits, 64 for the store and `usize` for the index within the store.
* In the C API we discard the 64 for the store and only have 64 for the `usize`, so the C API still uses 64-bit integers.

That would make the C API more unsafe than it is right now, but this change could be implemented backwards compatibly (even ABI-compatibly!). That makes me much less worried about the limits imposed here.

Reply:
And another option of course is some sort of compile-time configuration where the number of bits allocated to the index is a build time variable which specific embeddings could configure. This is so obscure, though, that I'd prefer no one else have to ever worry about this.

Reply:
Could you help me pinpoint where in the C API the 64-bit requirement is placed on `Stored<T>`?  I assume it's related to `wasmtime_extern_union` but I haven't been able to spot where the size assumption comes into play.

That said, this scheme makes sense to me assuming that requirement.

Reply:
The main requirement is that in the C API I would like to avoid types like `wasmtime_func_t` being owned types. Instead I'd like them to be POD-like with no destructor. For ergonomics they're also currently passed by-value everywhere in functions, e.g. returning `wasmtime_func_t` or taking `wasmtime_func_t func` or something like that. By-value structs always give me the willies (and I think concretely FFI interfaces like Python's `ctypes` doesn't support them) so that limited us to integers. The biggest integer for that was 64-bits, hence the 64-bit restriction.

There's other ways to change this though:

* With my idea above the 64-bits in the C API are just index bits, so nothing would be part of the store bits. Then Rust could use 128 bits and C could use 64.
* Another possibility is having the C API use 128-bit integers. This would probably show up as `struct wasmtime_func_t { uint64_t hi; uint64_t low; }` instead of a 128-bit integer, and then all APIs would take pointers to the struct (it'd never be returned by value)
* Alternatively we could go back to `wasmtime_func_t*` and require dtors,  but I would personally prefer to avoid this.

While I originally felt that it was kinda nice to have 64-bit types in the C API, I may be coming around to the 128-bit-size-struct idea. That keeps safety in both Rust & C (somewhat at least) and I think shouldn't make bindings really that much more difficult. I think the ergonomics of the C bindings will go down a bit but that's not really the end of the world per se.

Do you think it's worth these shenanigans to get 64-bit types? Or should we just go ahead and use 128-bits everywhere?

Reply:
Hm so having thought a bit more about this [this is the diff](https://gist.github.com/alexcrichton/25df779540182eefb8a833d0f9141160) of "what if all the types in the wasmtime API were 128-bit values". That actually looks relatively clean to me and makes me think that I should probably just change to that? The Go/Python bindings had no trouble updating to using that idiom either.

This also means that if we ever want to store more stuff in there it shouldn't really be a problem either?

Reply:
I think I would lean towards making that change so that we don't have to place limits on the number of stores that can be created (even if the limit seems reasonable today).

I looked over that commit and it looks good (minor spelling error on "belogns" in a few places, though).